src/main/java/nl/uu/cs/ape/models/templateFormulas/SLTLxTemplateFormula.java

```diff
@@ -615,11 +615,61 @@ public static String notConnectedModules(TaxonomyPredicate firstPredicate, Taxon
         return constraints.toString();
     }
 
+    /**
+     * Creates a CNF representation of the Constraint:<br>
+     * Do not repeat using the same tool.
+     * 
+     * @param predicate
+     * @param domainSetup
+     * @param moduleAutomaton
+     * @param typeAutomaton
+     * @param mappings
+     * @return
+     */
     public static String notRepeatModules(TaxonomyPredicate predicate, APEDomainSetup domainSetup,
             ModuleAutomaton moduleAutomaton,
             TypeAutomaton typeAutomaton, SATAtomMappings mappings) {
         StringBuilder constraints = new StringBuilder();
 
+        int automatonSize = moduleAutomaton.getAllStates().size();
+        for (int op1 = 0; op1 < automatonSize - 1; op1++) {
+            for (int op2 = op1 + 1; op2 < automatonSize; op2++) {
+
+                State firstModuleState = moduleAutomaton.get(op1);
+                State secondModuleState = moduleAutomaton.get(op2);
+
+                // filter all operations
+                domainSetup.getAllModules().getElementsFromSubTaxonomy(predicate).stream()
+                        .filter(x -> x.isSimplePredicate()).forEach(operation -> {
+
+                            constraints.append("-")
+                                    .append(mappings.add(operation, firstModuleState, AtomType.MODULE)).append(" ");
+                            constraints.append("-"
+                                    + mappings.add(operation, secondModuleState, AtomType.MODULE)
+                                    + " 0\n");
+                        });
+
+            }
+        }
+
+        return constraints.toString();
+    }
+
+    /**
+     * Creates a CNF representation of the Constraint:<br>
+     * The same tools that belong to the sub-taxonomy should not be connected.
+     * 
+     * @param predicate       - Root predicate of the sub-taxonomy.
+     * @param domainSetup
+     * @param moduleAutomaton
+     * @param typeAutomaton
+     * @param mappings
+     * @return
+     */
+    public static String notConnectModules(TaxonomyPredicate predicate, APEDomainSetup domainSetup,
+            ModuleAutomaton moduleAutomaton,
+            TypeAutomaton typeAutomaton, SATAtomMappings mappings) {
+        StringBuilder constraints = new StringBuilder();
         int automatonSize = moduleAutomaton.getAllStates().size();
         for (int op1 = 0; op1 < automatonSize - 1; op1++) {

```

Comment:
I see the same pattern of pairing the ops as in lines 635ff here. Just a refactoring proposal: You could write a function producing you int pairs over which you then only have to iterate.